### PR TITLE
Enables bfloat16 x [float16, complex64, complex128] type promotion

### PR DESCRIPTION
**Pull request by mruberry**
Implements bfloat16 type promotion consistent with JAX (see https://jax.readthedocs.io/en/latest/type_promotion.html), addressing issue https://github.com/pytorch/pytorch/issues/43049.

- bfloat16 x float16 -> float32
- bfloat16 x complex64 -> complex64
- bfloat16 x complex128 -> complex128

Existing tests, after updates, are sufficient to validate the new behavior.

cc @xuhdev

### CODE REVIEW

File: test/test_type_promotion.py [hunk 1]
```diff
@@ -174,38 +174,27 @@ def test_bfloat16(self, device):
             self.assertEqual((bf + scalar).dtype, torch.bfloat16)
             self.assertEqual((scalar + bf).dtype, torch.bfloat16)
             self.assertEqual(scalar + bf, bf + scalar)
-        with self.assertRaises(RuntimeError):
```
Comment (gchanan) [inline test/test_type_promotion.py:177]: can you add a complex scalar to the for loop above?  That seems to capture the intent of this test.

File: test/test_type_promotion.py [hunk 2]
```diff
@@ -174,38 +174,27 @@ def test_bfloat16(self, device):
...
-                with self.assertRaises(RuntimeError):
-                    t + bf
-                with self.assertRaises(RuntimeError):
-                    torch.promote_types(dtype, torch.bfloat16)
-                with self.assertRaises(RuntimeError):
-                    torch.promote_types(torch.bfloat16, dtype)
+            self.assertEqual(bf + t, t + bf)
+            if dtype in (torch.float16, torch.float32, torch.float64, torch.cfloat, torch.cdouble):
+                # Handles bfloat16 x float16 -> float32 promotion
+                expected_dtype = dtype if dtype != torch.half else torch.float32
```
Comment (gchanan) [inline test/test_type_promotion.py:184]: can you rationalize?

**Review by gchanan — APPROVED** [review]

File: test/test_type_promotion.py (hunk 1 above)
Comment (mruberry) [reply test/test_type_promotion.py:177]: Good idea. I also simplified the loop.

File: test/test_type_promotion.py (hunk 2 above)
Comment (mruberry) [reply test/test_type_promotion.py:184]: Rationalized.

**Pull request closed without merging**